dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java

```diff
@@ -1383,14 +1452,21 @@ synchronized String evaluateInterval() {
             int poolAdjustment = 0;
             if (growScore >= growScoreFilterLevel && (growScore - growShrinkDiffFilter) > shrinkScore) {
                 poolAdjustment = poolIncrement;
+                // with available cpu and a persistently deep queue, grow the pool by more than usual
+                if (!cpuHigh && !systemCpuNA && deepQueue) {
+                    if (consecutiveDeepQueueCycles > consecutiveDeepQueueThreshold) {
+                        poolAdjustment = Math.min(poolAdjustment * deepQueuePoolIncrementMultiple, maxThreads - poolSize);
```

Comment:
I had similar thoughts, and concluded that if we need a lot of threads, the increment has to be large. Thought about doing something arbitrary like adding a thread for every task in queue, but the historical data collection should work better if we stay on a finite set of pool sizes.

Reply:
Well ... we don't execute the controller logic again until all threads are added - there's logic that skips the cycle if the thread count does not reflect the most recent adjustment. So each cycle is a one-time decision. 

The logic for adding a lot of threads is that the queue represents the current state of activity: the current thread count is not keeping up with incoming load, and if load continues at this rate more threads will be needed (assuming CPU is available).

If we come around on the next cycle and the queue is empty, then we'll go down that path, which will tend to reduce the pool size ... and if the queue is empty, then life is easy, and having more threads than we need at the moment is not a big deal.

I thought about handling load surges by shortening the controller cycle rather than increasing the poolIncrement. That would require creating a "deepQueueAndCpuAvailable" state in the controller, analogous to the "poolHung" state. In that model we might chop the interval in half or quarter, so that we are checking on things more often, and (if needed) can add lots of threads in smaller increments. That would be a little more complicated, but I'm still open to the idea, if you think it might be preferable.